allennlp/commands/predict.py

```diff
@@ -151,12 +151,22 @@ def _maybe_print_to_console_and_file(self,
             self._output_file.write(prediction)
 
     def _get_json_data(self) -> Iterator[JsonDict]:
-        for line in open(self._input_file):
-            if not line.isspace():
-                yield self._predictor.load_line(line)
+        try:
+            if self._input_file == "-":
+                source = sys.stdin
+            else:
+                source = open(self._input_file)
+
+            for line in source:
+                if not line.isspace():
+                    yield self._predictor.load_line(line)
+        finally:
+            source.close()
```

Comment:
I feel like this would be more readable if we didn't try to handle both cases at the same time. e.g.

```
            if self._input_file == "-":
                # read from stdin
                for line in sys.stdin:
                    if not line.isspace():
                        yield self._predictor.load_line(line)
            else:
                # read from file
                with open(self._input_file) as source:
                    for line in source:
                        if not line.isspace():
                            yield self._predictor.load_line(line)
```